Fix metrics

degaydorus opened the pull request:
This PR allows for clients to pass in a metric prefix.  It first supported this by allowing cache registration to be done on the client side, using cache builder methods in mt-dynamo.  To maintain encapsulation, this was modified to allow a client to pass in a metric prefix when creating a `CachingAmazonDynamoDbStreams` instance.
- added tests for monitoring caches created in `CachingAmazonDynamoDbStreams`
- added builder for describe stream cache to use in tests (previously exposed getter)

busjaeger commented on src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java:
```diff
@@ -217,6 +221,17 @@ public Builder withMaxIteratorCacheSize(int maxIteratorCacheSize) {
+        public Builder withDescribeStreamCache(Cache describeStreamCache) {
+            this.describeStreamCache = describeStreamCache;
+            return this;
+        }
```
Do you still need to expose this?

degaydorus replied:
https://github.com/salesforce/mt-dynamo/pull/509/files#diff-2a4b73d2ee26b7ce82d49248a4931568 `MtAmazonDynamoDbStreamsBaseTest ` uses it in the test.  Before there was a getter for the test, in the past we discussed adding a builder instead.  In both options (getter or builder method), the test need a way to get to the cache.

busjaeger commented on src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java:
```diff
@@ -720,29 +739,26 @@ private static String toShortString(GetRecordsResult result) {
-    // TODO: introduce builder/constructor variant that allows passing in cache reference
-    public Cache<String, DescribeStreamResult> getDescribeStreamCache() {
-        return describeStreamCache;
+        final String prefix = StringUtils.isNullOrEmpty(metricPrefix) ? cn : cn + "." + metricPrefix;
```
Just curious: why did you decide to go with this option vs injecting the prefix via a `MeterRegistry` in ZOS?

degaydorus replied:
I wanted to talk to you about that, I did try implementing `MeterRegistry`, and had a couple concerns.  See below are some examples, in creating a timer, I chose to call the method that takes tags, but there are other timer methods (one that takes strings), but there wasn't an option to pass in distributionStatisticConfig for example.  Most of the registries that implement `MeterRegistry` have separate classes for each type, i.e. Composite counter.  I think we just want to add the prefix and not have custom metric types, basically wasn't sure how in depth we wanted to go with the implementation.

    @Override
    protected Timer newTimer(Id id, DistributionStatisticConfig distributionStatisticConfig,
                             PauseDetector pauseDetector) {
        return meterRegistry.timer(prefix + "." + id.getName(), id.getTags());
    }

    @Override
    protected DistributionSummary newDistributionSummary(Id id, DistributionStatisticConfig distributionStatisticConfig,
                                                         double scale) {
        return meterRegistry.summary(id.getName(), id.getTags());
    }

degaydorus commented:
> As discussed could considering taking out cache builder method.

I removed the builder method for describeStreamCache, added the getter only for `CachingAmazonDynamoDbStreams` because checking the content of what's added in the cache is still necessary for some of those tests.  The getter has access scope reduced to non-modifier instead of public.

`MtAmazonDynamoDbStreamsBaseTest` just checks the meterRegistry because that's all that's needed at that level (not the contents of what's cached).
